jaxrs-spec/src/main/asciidoc/chapters/providers/_lifecycle_and_environment.adoc

```diff
@@ -29,9 +29,20 @@ all of the lifecycle options provided by that framework.
 ==== Automatic Discovery
 
 The annotation `@Provider` is used by a JAX-RS runtime to automatically
-discover provider classes via mechanisms such as class scanning. A
-JAX-RS implementation that supports automatic discovery of classes MUST
-process only those classes that are annotated with `@Provider`.
+discover provider classes via mechanisms such as class scanning.
+
+[[extensions]]
+Providers MAY implement the `Extension` interface to declare it as an
+extension of the runtime.
+Compatible products MUST automatically register a class in
+`jakarta.ws.rs.core.Configuration` runtime configurations if the
+class is a _provider class_ for the `Extension` _service type_ according
+to the rules set out in the description of `ServiceLoader`.
+
+A JAX-RS implementation that supports automatic discovery of classes MUST
+process only those classes that are annotated with `@Provider` or are
```

Comment:
@chkal You added an *ONLY* restriction here, which in fact would prevent Jersey to further process its *proprietary* annotation. Is this really what you want? And is this agreed by @jansupol ?

Reply:
I believe two principles are mixed here together. The automatic discovery, i.e. classpath scanning, and JDK service loading mechanism. The chapter is about Automatic Discovery and it should not be mixed with the Extensions, which are not loaded by a classpath scanning. If the meaning would be as @mkarg suggests, any service loaded by an implementation via ServiceLoader would be in trouble. 

A new chapter regarding Extensions without a connection to the Automatic Discovery seems to be more appropriate, I.e. keep section 4.1.1 Autodiscovery as is, and add
```
==== Extensions

Providers MAY implement the `Extension` interface to declare it as an extension of the runtime. 
Compatible products MUST automatically register a class in `jakarta.ws.rs.core.Configuration` 
runtime configurations if the class is a _provider class_ for the `Extension` _service type_ according to 
the rules set out in the description of `ServiceLoader`.
```

Still, I believe the Extensions are going to be a source of trouble where one app will affect the second with their Extension and it should be possible to opt out the Extensions by an Application setting. The similar `AutoDiscoverable` feature in Jersey was meant for Jersey modules, and for the integration layers, but not for the end users.

Reply:
Even better yet, the Auto Discovery needs to be turned on (either by a missing Application or by returning an empty collection by both `Application.getClasses` and `Application.getSingletons`), and similarly, I'd prefer the Extension support to be turned on before it starts working, either by a missing Application or by annotating the existing Application with a new annotation.

Reply:
Opting out: We already discussed this and agreed that this a new opt-out property would be good, but this has nothing to this with *this* discussion, as it is solely about documenting the feature *as it already exists* in `master` and that everybody wanting new feature must file his own PR.

Reply:
> @chkal You added an _ONLY_ restriction here, which in fact would prevent Jersey to further process its _proprietary_ annotation. Is this really what you want? And is this agreed by @jansupol ?

Good point. The `only` restriction was already present before, but just referred to `@Provider`. So I could argue that Jersey (and most other implementations) are already violating this part of the spec with their proprietary mechanism of registering providers. No?

> I believe two principles are mixed here together. The automatic discovery, i.e. classpath scanning, and JDK service loading mechanism.

IMO using `Configurable#register()` is *manual* and everything else (`@Provider` and `Extension`) is *automatic* discovery, as it doesn't require explicit calls to `register`.

However, I would also be fine with moving the `Extension` part to a separate section if everybody agrees that this would be better.